File: milk4.cpp
/*
ID: wrwwctb1
LANG: C++14
TASK: milk4
*/

#include <stdio.h>
#include <stdlib.h>
#define MAXQ 20001
#define MAXP 101
#define INF  0x7fffffff
#include <vector>
int betterSequence (int a, int b, int l[], int n[]);
int compare (const void *a, const void *b);

int Q, P;
int pail[MAXP];
int s[MAXQ];
int l[MAXQ];
int n[MAXQ];

int main (){
    FILE *fin = fopen ("milk4.in", "r");
    fscanf (fin, "%d %d", &Q, &P);
    for (int i = 1; i <= P; i++)
	    fscanf (fin, "%d", &(pail[i]));
    fclose (fin);
    qsort (pail+1, P, sizeof (pail[0]), compare);


    for (int i = 0; i <= Q; i++) {
        s[i] = INF;
        l[i] = 0;
        n[i] = 0;
    }

    s[0] = 0;

    for (int i = 1; i <= P; i++) {
        int ts[MAXQ], tl[MAXQ], tn[MAXQ];

        for (int q = 0; q <= Q; q++) {
            ts[q] = s[q];
            tl[q] = l[q];
            tn[q] = n[q];
        }

        for (int q = pail[i]; q <= Q; q++) {
            int prevq = q - pail[i];

            if (ts[prevq] < INF){
                if (tl[prevq] == pail[i]) {
                    ts[q] = ts[prevq];
                    tl[q] = pail[i];
                    tn[q] = tn[prevq] + 1;
                }else{
                    ts[q] = ts[prevq] + 1;
                    tl[q] = pail[i];
                    tn[q] = 1;
                }

                if (s[prevq] < INF &&
                    (s[prevq] + 1 < ts[q] ||
                     (s[prevq] + 1 == ts[q] &&
                      betterSequence (prevq, q - tn[q] * pail[i], l, n)
                     )
                    )
                   ) {
                    ts[q] = s[prevq] + 1;
                    tl[q] = pail[i];
                    tn[q] = 1;
                }
            }
        }

        for (int q = pail[i]; q <= Q; q++)
            if (ts[q] <= s[q]) {
                s[q] = ts[q];
                l[q] = tl[q];
                n[q] = tn[q];
            }
    }


    int q = Q;
    std::vector<int> traced;
    while (q > 0){
        printf(" %d %d %d\n", q, l[q], n[q]);
	    traced.push_back(l[q]);
	    q -= l[q] * n[q];
    }


    FILE *fout = fopen ("milk4.out", "w");
    fprintf (fout, "%d", s[Q]);

    for (int i = 0; i < s[Q]; i++)
	    fprintf (fout, " %d", traced[i]);
    fprintf (fout, "\n");
    fclose (fin);

    return 0;
}



int betterSequence (int a, int b, int l[], int n[]){
    if (a == b)
        return 0;
    while (a && b) {
        if (l[a] < l[b])
            return 1;
        if (l[a] > l[b])
            return 0;

        a -= n[a] * l[a];
        b -= n[b] * l[b];
    }

    if (a)
	    return 0;
    return 1;
}

int compare (const void *a, const void *b){
    return *(int *) b - *(int *) a;
}

File: milk4.py
"""
ID: wrwwctb1
LANG: PYTHON3
TASK: milk4
5.3.2

key idea
unbounded knapsack with tricky solution selection

so far the most brutal

point 1/2:
each s[q] only depends on itself and s[q-pail[i]]
it's a lot more subtle to trace the required optimal set
see note

point 2/2:
drastic speed difference (>>10) between milk4.cpp and milk4.py for milk46.in
milk4.py and milk4.cpp use exact same algo. only latter can pass
milk4.py1 uses search and dp. a c++ translation would've passed
"""

import os, sys, re, time
from collections import Counter, deque, defaultdict
from queue import Queue
from copy import copy, deepcopy
from itertools import combinations, permutations, accumulate, \
                      combinations_with_replacement
from functools import lru_cache, cmp_to_key, partial as functools_partial
from heapq import heappush, heappop, nlargest, nsmallest
from bisect import bisect_left, bisect_right
from math import ceil, floor, factorial, gcd, modf, log, log2, log10, sqrt, \
         pi, sin, cos, tan, asin, acos, atan, atan2, hypot, erf, erfc, inf, nan
# sys.setrecursionlimit(5782)
# python -m cProfile -s time ha.py

print = functools_partial(print, flush=True)
input = lambda: fin.readline().strip('\n')

def printwrite(string):
    print(string)
    fout.write(str(string) + '\n')

filename = 'milk4'
fin = open(filename + '.in', 'r')
fout = open(filename + '.out', 'w')










def show():
#    ' '.join(['%4s'] * (Q+1)) % tuple(map(str, s))
    print('   q    s    l   n')
    for q in range(Q+1):
        print('%4s %4s %4s %4s' % \
              (str(q), str(s[q]), str(l[q]), str(n[q])))
    print('pails', pail)

def track(qa, qb, l, n):
    if qa == qb:
        return False
    while qa and qb:
        if l[qa] < l[qb]:
            return True
        elif l[qa] > l[qb]:
            return False
        else:
            qa -= n[qa] * l[qa]
            qb -= n[qb] * l[qb]
    if qa == 0:
        return False
    else:
        return True

Q = int(input())
P = int(input())
pail = []
for p in range(P):
    pail.append(int(input()))

pail = [None] + sorted(set(pail), reverse=True)


s = [inf] * (Q + 1) # s[q] = size of min set pail for q
l = [0]   * (Q + 1) # l[q] = last pail used
n = [0]   * (Q + 1) # n[q] = num of times last pail used

s[0] = 0 # no pail is needed for q==0

for i in range(1, len(pail)): # run through pails
    paili = pail[i]

    ts = copy(s)
    tl = copy(l)
    tn = copy(n)

    for q in range(1, Q+1):
        if q >= paili:
            prevq = q - paili

            if ts[prevq] < inf: # prevq reachable from top
                if tl[prevq] == paili:
                    ts[q] = ts[prevq]
                    tl[q] = paili
                    tn[q] = tn[prevq] + 1
                else:
                    ts[q] = ts[prevq] + 1
                    tl[q] = paili
                    tn[q] = 1

                if s[prevq] < inf: # prevq reachable from left

                    # is left better?
                    if (s[prevq] + 1 < ts[q] or # strictly better
                        (s[prevq] + 1 == ts[q] and # tie
                         track(prevq, q - tn[q] * paili, l, n))):
                         # ^ compare two branches:
                         # (1) prevq reached from left, paili not included
                         # (2) topmost entry in column i, paili not included

                        ts[q] = s[prevq] + 1
                        tl[q] = paili
                        tn[q] = 1

    # is q reachable from left? is it better? update s l n
    for q in range(1, Q+1):
        if ts[q] <= s[q]:
            s[q] = ts[q]
            l[q] = tl[q]
            n[q] = tn[q]
#    show()

q = Q
traced = []
while q > 0:
    print(q, l[q], n[q])
    traced.append(l[q])
    q -= l[q] * n[q]

printwrite(' '.join(['%d'] * (s[-1] + 1)) % tuple([s[-1]] + traced))

fin.close()
fout.close()
del print, input

